# SHEETAL PATIL | DATA PREPROCESSING AND FEATURE ENGINEERING (EDA2)

This analysis outlines the data exploration, preprocessing, feature engineering, and selection processes performed on the Adult dataset.

In [40]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder
from sklearn.ensemble import IsolationForest
from sklearn.feature_selection import mutual_info_classif

In [3]:
# Load the dataset
df = pd.read_csv("D:\\Work\\ExcelRCource\\Assignments\\12 EDA2\\adult_with_headers.csv")
df

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [5]:
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [6]:
df.describe(include= "object")

,workclass,education,marital_status,occupation,relationship,race,sex,native_country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


In [7]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [9]:
# Check for '?' which is common in this dataset
question_mark_counts = (df == ' ?').sum() # Notice the space before ? in some versions
question_mark_counts

age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     583
income               0
dtype: int64

**Basic Exploration**
The dataset contains 32,561 entries and 15 columns.

**Numerical Features:** age, fnlwgt, education_num, capital_gain, capital_loss, hours_per_week.

**Categorical Features:** workclass, education, marital_status, occupation, relationship, race, sex, native_country, income.

**Missing Values:** While standard .isnull() returned zero, the dataset uses ' ?' to represent missing values in workclass (1,836), occupation (1,843), and native_country (583).

### 1. Data Preprocessing 

In [12]:
# Replace ' ?' with NaN and then impute with mode
df.replace(' ?', np.nan, inplace=True)
for col in ['workclass', 'occupation', 'native_country']:
    df[col].fillna(df[col].mode()[0], inplace=True)

C:\Users\Admin\AppData\Local\Temp\ipykernel_12500\213279794.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].mode()[0], inplace=True)


In [13]:
(df == ' ?').sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

**Handling Missing Values**
Missing values were imputed using the Mode (most frequent value) for each respective categorical column. This ensures that the dataset remains complete while preserving the original distribution as much as possible.

In [14]:
# Numerical columns
num_cols = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

In [15]:
# Standard Scaling
scaler_std = StandardScaler()
df_std = df.copy()
df_std[num_cols] = scaler_std.fit_transform(df[num_cols])

In [16]:
# Min-Max Scaling
scaler_minmax = MinMaxScaler()
df_minmax = df.copy()
df_minmax[num_cols] = scaler_minmax.fit_transform(df[num_cols])

**Feature Scaling**
Two primary scaling techniques were applied to numerical features:

**Standard Scaling ($Z$-score normalization):** Centers the data (mean = 0) and scales to unit variance (std = 1).\
**Min-Max Scaling:** Rescales the data to a fixed range, typically $[0, 1]$.

Scenarios for Each Technique:

**Standard Scaling:** Preferred for algorithms that assume normally distributed data (e.g., Linear Regression, Logistic Regression) and is more robust to outliers compared to Min-Max.

**Min-Max Scaling:** Preferred for algorithms that do not assume any distribution and require features to be on the same scale (e.g., K-Nearest Neighbors, Neural Networks). It is also essential when data must be bounded (e.g., image processing).

In [17]:
df.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


Feature-------|----Standard Scaled (Mean ≈ 0)---|---Min-Max Scaled (Range 0-1)\
Age-----------|------0.030---------------------|-------------0.301\
Hours per Week|------−0.035-----------------|----------0.398

### 2. Encoding

**Encoding Techniques**\
Categorical variables were encoded based on their cardinality (number of unique categories):\
**One-Hot Encoding (OHE):** Applied to sex and income ($< 5$ categories).\
**Label Encoding (LE):** Applied to workclass, education, marital_status, occupation, relationship, race, and native_country ($\ge 5$ categories).

In [18]:
cat_cols = df.select_dtypes(include=['object']).columns
unique_counts = {col: df[col].nunique() for col in cat_cols}
print("Unique counts for categorical columns:", unique_counts)

Unique counts for categorical columns: {'workclass': 8, 'education': 16, 'marital_status': 7, 'occupation': 14, 'relationship': 6, 'race': 5, 'sex': 2, 'native_country': 41, 'income': 2}


In [19]:
# One-Hot Encoding for < 5 categories
# Label Encoding for >= 5 categories
df_encoded = df.copy()
label_encoder = LabelEncoder()

In [21]:
ohe_cols = [col for col in cat_cols if df[col].nunique() < 5]
le_cols = [col for col in cat_cols if df[col].nunique() >= 5]

print("Columns for OHE (<5):", ohe_cols)
print("Columns for LE (>=5):", le_cols)

Columns for OHE (<5): ['sex', 'income']
Columns for LE (>=5): ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'native_country']


In [27]:
# Apply One-Hot
df_encoded = pd.get_dummies(df_encoded, columns=ohe_cols)

KeyError: "None of [Index(['sex', 'income'], dtype='object')] are in the [columns]"

In [25]:
# Apply Label Encoding
for col in le_cols:
    df_encoded[col] = label_encoder.fit_transform(df_encoded[col])

In [26]:
df_encoded.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K
0,39,6,77516,9,13,4,0,1,4,2174,0,40,38,False,True,True,False
1,50,5,83311,9,13,2,3,0,4,0,0,13,38,False,True,True,False
2,38,3,215646,11,9,0,5,1,4,0,0,40,38,False,True,True,False
3,53,3,234721,1,7,2,5,0,2,0,0,40,38,False,True,True,False
4,28,3,338409,9,13,2,9,5,2,0,0,40,4,True,False,True,False


### 3. Feature Engineering

**Feature Engineering**\
To enhance model predictive power, the following transformations and features were created:\
**1)capital_diff ($Capital\ Gain - Capital\ Loss$):**\
Rationale: Instead of looking at gains and losses separately, this captures the net investment success of an individual, which is a stronger indicator of wealth.\
**2)age_edu_interaction ($Age \times Education\ Num$):**\
Rationale: The impact of education on income often compounds with age. A highly educated older individual is statistically more likely to be in a high-income bracket than a highly educated younger individual.\
**3)Log Transformation (capital_gain_log):**\
Choice: Applied to capital_gain because it is highly right-skewed with many zeros and extreme outliers.\
Justification: Transformation using $\log(1+x)$ compresses the range of extreme values, making the distribution more Gaussian-like and helping linear models perform better.

In [28]:
# 1. Net Capital
df_encoded['capital_diff'] = df_encoded['capital_gain'] - df_encoded['capital_loss']

In [29]:
# 2. Education-Age Interaction or Binary Gender
df_encoded['age_edu_interaction'] = df_encoded['age'] * df_encoded['education_num']

In [30]:
# Log transformation for skewed feature (capital_gain)
# Log(1+x) to handle zeros
df_encoded['capital_gain_log'] = np.log1p(df_encoded['capital_gain'])

In [31]:
df_encoded.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K,capital_diff,age_edu_interaction,capital_gain_log
0,39,6,77516,9,13,4,0,1,4,2174,0,40,38,False,True,True,False,2174,507,7.684784
1,50,5,83311,9,13,2,3,0,4,0,0,13,38,False,True,True,False,0,650,0.000000
2,38,3,215646,11,9,0,5,1,4,0,0,40,38,False,True,True,False,0,342,0.000000
3,53,3,234721,1,7,2,5,0,2,0,0,40,38,False,True,True,False,0,371,0.000000
4,28,3,338409,9,13,2,9,5,2,0,0,40,4,True,False,True,False,0,364,0.000000


### 4. Feature Selection

**Feature Selection and Outlier Detection**\
**Isolation Forest for Outliers**\
The Isolation Forest algorithm was used to identify anomalies. Approximately 1,628 outliers ($5\%$) were detected and removed.\
**Impact of Outliers:** Outliers can disproportionately influence the mean and variance, leading to biased scaling. In modeling, they can pull the decision boundary towards noise, resulting in poor generalization and overfitting.

In [32]:
# Isolation Forest for outliers
iso = IsolationForest(contamination=0.05, random_state=42)
outliers = iso.fit_predict(df_encoded.select_dtypes(include=[np.number]))
df_no_outliers = df_encoded[outliers == 1]

print(f"Removed {np.sum(outliers == -1)} outliers.")

Removed 1628 outliers.


Since the ppscore library is a specialized tool, Mutual Information (MI) was used as a proxy to assess relationships.\
**Correlation Matrix (Pearson):** Measures linear relationships. It showed a strong correlation between education_num and income ($0.335$). However, it fails to capture non-linear dependencies.\
**PPS / Mutual Information:** MI captures any type of statistical dependency (linear or non-linear).\
**Findings:** MI revealed that marital_status and relationship have higher predictive power than age or hours_per_week, despite age having a higher linear correlation score.

In [34]:
# Correlation Matrix
corr_matrix = df_encoded.corr()
corr_matrix

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K,capital_diff,age_edu_interaction,capital_gain_log
age,1.000000,0.040504,-0.076646,-0.010508,0.036527,-0.266288,0.001739,-0.263698,0.028718,0.077674,0.057775,0.068756,-0.000270,-0.088832,0.088832,-0.234037,0.234037,0.074284,0.762507,0.124183
workclass,0.040504,1.000000,-0.024338,0.004874,0.003536,-0.020468,0.007110,-0.057947,0.048350,0.031505,0.002644,0.042199,-0.001625,-0.071584,0.071584,-0.002693,0.002693,0.031261,0.033176,0.015338
fnlwgt,-0.076646,-0.024338,1.000000,-0.028145,-0.043195,0.028153,0.000188,0.008931,-0.021291,0.000432,-0.010252,-0.018768,-0.063286,-0.026858,0.026858,0.009463,-0.009463,0.000988,-0.078272,-0.004414
education,-0.010508,0.004874,-0.028145,1.000000,0.359153,-0.038407,-0.041279,-0.010876,0.014131,0.030046,0.016746,0.055510,0.076060,0.027356,-0.027356,-0.079317,0.079317,0.029039,0.202574,0.024955
education_num,0.036527,0.003536,-0.043195,0.359153,1.000000,-0.069304,0.070954,-0.094153,0.031838,0.122630,0.079923,0.148123,0.088894,-0.012280,0.012280,-0.335154,0.335154,0.117891,0.640195,0.129135
marital_status,-0.266288,-0.020468,0.028153,-0.038407,-0.069304,1.000000,0.034962,0.185451,-0.068013,-0.043393,-0.034187,-0.190519,-0.021278,0.129314,-0.129314,0.199307,-0.199307,-0.041395,-0.246621,-0.066595
occupation,0.001739,0.007110,0.000188,-0.041279,0.070954,0.034962,1.000000,-0.037451,-0.004839,0.018021,0.009680,-0.012879,-0.002217,-0.047461,0.047461,-0.034625,0.034625,0.017437,0.043048,0.010884
relationship,-0.263698,-0.057947,0.008931,-0.010876,-0.094153,0.185451,-0.037451,1.000000,-0.116055,-0.057919,-0.061062,-0.248974,-0.010712,0.582454,-0.582454,0.250918,-0.250918,-0.054413,-0.249747,-0.083402
race,0.028718,0.048350,-0.021291,0.014131,0.031838,-0.068013,-0.004839,-0.116055,1.000000,0.011145,0.018899,0.041910,0.116529,-0.087204,0.087204,-0.071846,0.071846,0.010081,0.047529,0.024089
capital_gain,0.077674,0.031505,0.000432,0.030046,0.122630,-0.043393,0.018021,-0.057919,0.011145,1.000000,-0.031615,0.078409,0.008819,-0.048480,0.048480,-0.223329,0.223329,0.998521,0.143722,0.564520


In [35]:
# Check for ppscore
try:
    import ppscore as pps
    pps_matrix = pps.matrix(df_encoded)
    pps_available = True
except ImportError:
    pps_available = False

print("PPS Available:", pps_available)

PPS Available: True


Predictive Power Score (PPS) vs. Correlation Matrix\
**Comparison:** The correlation matrix is excellent for identifying simple linear trends, whereas PPS/MI is superior for identifying complex patterns that modern non-linear models (like Random Forests or XGBoost) can exploit.

In [37]:
# Prepare summary for output
print("\nFirst few rows of encoded and engineered dataframe:")
df_encoded.head()


First few rows of encoded and engineered dataframe:


,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K,capital_diff,age_edu_interaction,capital_gain_log
0,39,6,77516,9,13,4,0,1,4,2174,0,40,38,False,True,True,False,2174,507,7.684784
1,50,5,83311,9,13,2,3,0,4,0,0,13,38,False,True,True,False,0,650,0.000000
2,38,3,215646,11,9,0,5,1,4,0,0,40,38,False,True,True,False,0,342,0.000000
3,53,3,234721,1,7,2,5,0,2,0,0,40,38,False,True,True,False,0,371,0.000000
4,28,3,338409,9,13,2,9,5,2,0,0,40,4,True,False,True,False,0,364,0.000000


### Compare its findings with the correlation matrix

In [38]:
# Define target for MI comparison
# Let's use 'income_ >50K' as the target
target = 'income_ >50K'
X = df_encoded.drop(['income_ <=50K', 'income_ >50K'], axis=1)
y = df_encoded[target]

In [41]:
# Mutual Information scores
mi_scores = mutual_info_classif(X, y, random_state=42)
mi_series = pd.Series(mi_scores, index=X.columns).sort_values(ascending=False)

In [42]:
# Correlation with target
corr_with_target = df_encoded.corr()[target].drop(['income_ <=50K', 'income_ >50K']).sort_values(ascending=False)

In [45]:
print("Mutual Information Scores with Target:")
print(mi_series)

Mutual Information Scores with Target:
capital_diff           0.120213
marital_status         0.112612
relationship           0.112461
age_edu_interaction    0.108820
capital_gain_log       0.082944
capital_gain           0.078196
education              0.067264
age                    0.067081
education_num          0.061205
occupation             0.051748
hours_per_week         0.042710
fnlwgt                 0.033464
capital_loss           0.032686
sex_ Male              0.026494
sex_ Female            0.025476
native_country         0.011729
workclass              0.011098
race                   0.010880
dtype: float64


In [48]:
print("\nCorrelation with Target:")
corr_with_target


Correlation with Target:


age_edu_interaction    0.393086
education_num          0.335154
capital_gain_log       0.289462
age                    0.234037
hours_per_week         0.229689
capital_gain           0.223329
sex_ Male              0.215980
capital_diff           0.214428
capital_loss           0.150526
education              0.079317
race                   0.071846
occupation             0.034625
native_country         0.023058
workclass              0.002693
fnlwgt                -0.009463
marital_status        -0.199307
sex_ Female           -0.215980
relationship          -0.250918
Name: income_ >50K, dtype: float64

In [47]:
# Correlation Matrix (subset for readability)
corr_matrix = df_encoded.corr()
print("\nCorrelation Matrix (Partial):")
corr_matrix.iloc[:10, :10]


Correlation Matrix (Partial):


,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain
age,1.000000,0.040504,-0.076646,-0.010508,0.036527,-0.266288,0.001739,-0.263698,0.028718,0.077674
workclass,0.040504,1.000000,-0.024338,0.004874,0.003536,-0.020468,0.007110,-0.057947,0.048350,0.031505
fnlwgt,-0.076646,-0.024338,1.000000,-0.028145,-0.043195,0.028153,0.000188,0.008931,-0.021291,0.000432
education,-0.010508,0.004874,-0.028145,1.000000,0.359153,-0.038407,-0.041279,-0.010876,0.014131,0.030046
education_num,0.036527,0.003536,-0.043195,0.359153,1.000000,-0.069304,0.070954,-0.094153,0.031838,0.122630
marital_status,-0.266288,-0.020468,0.028153,-0.038407,-0.069304,1.000000,0.034962,0.185451,-0.068013,-0.043393
occupation,0.001739,0.007110,0.000188,-0.041279,0.070954,0.034962,1.000000,-0.037451,-0.004839,0.018021
relationship,-0.263698,-0.057947,0.008931,-0.010876,-0.094153,0.185451,-0.037451,1.000000,-0.116055,-0.057919
race,0.028718,0.048350,-0.021291,0.014131,0.031838,-0.068013,-0.004839,-0.116055,1.000000,0.011145
capital_gain,0.077674,0.031505,0.000432,0.030046,0.122630,-0.043393,0.018021,-0.057919,0.011145,1.000000
